test/test_thread_utils.cpp

```diff
@@ -6,35 +6,85 @@
 
 namespace mamba
 {
-    TEST(thread_utils, interrupt)
+    namespace
     {
-#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
+        std::mutex res_mutex;
         int res = 0;
+    }
+#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
+    void test_interruption_guard(bool interrupt)
+    {
         // Ensures the compiler doe snot optimize away Context::instance()
         std::string current_command = Context::instance().current_command;
         EXPECT_EQ(current_command, "mamba");
-
         Console::instance().init_multi_progress();
         {
-            interruption_guard g([&res]() {
+            interruption_guard g([]() {
                 // Test for double free (segfault if that happends)
+                std::cout << "Interruption guard is interrupting" << std::endl;
                 Console::instance().init_multi_progress();
-                res = -1;
+                {
+                    std::unique_lock<std::mutex> lk(res_mutex);
+                    res -= 100;
+                }
+                reset_sig_interrupted();
             });
 
             for (size_t i = 0; i < 5; ++i)
             {
-                thread t([&res]() {
-                    ++res;
+                mamba::thread t([]() {
+                    {
+                        std::unique_lock<std::mutex> lk(res_mutex);
+                        ++res;
+                    }
                     std::this_thread::sleep_for(std::chrono::milliseconds(300));
                 });
                 t.detach();
             }
-
-            pthread_kill(get_cleaning_thread_id(), SIGINT);
-            std::this_thread::sleep_for(std::chrono::seconds(1));
+            if (interrupt)
+            {
+                pthread_kill(get_signal_receiver_thread_id(), SIGINT);
+                // sleep before resetting sig_interrupted
+                std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            }
         }
-        EXPECT_EQ(res, -1);
-#endif
     }
+
+    TEST(thread_utils, interrupt)
+    {
+        res = 0;
+        test_interruption_guard(true);
+        EXPECT_EQ(res, -95);
+    }
+
+    TEST(thread_utils, no_interrupt)
+    {
+        res = 0;
+        test_interruption_guard(false);
+        wait_for_all_threads();
```

Comment:
This is a bit annoying, this should be handled automatically in the interruption mechanism. Otherwise, the main thread needs to wait for the parallel task to complete before it resumes.

Maybe the `wait_for_all_threads` should always be called in the `interruption_guard` destructor?

Reply:
it is handled when we do have an interruption (but not in the `false` case).